### ISSUE TITLE
Introduce an interception mechanism

### ISSUE THREAD

**Issue by lehmamic** [angular/angular#2893]
It would be nice to have an interception mechanism similar to the $http interception from angular.js, but based on any services working together with the injection api. This would allow to introduce service behaviors based on aspect oriented programming, for example authorization, exception handling, logging etc.

As a reference I would like to mention the Unity dependency injection framework for .Net:
https://msdn.microsoft.com/en-us/library/dn178466(v=pandp.30).aspx

I could imagine that working with interceptors looks like following snippets:

Interceptor implementation
```javascript
class MyInterceptionBehavior implements InterceptionBehavior {
	intercept(input:FunctionInvocation, getNext:GetNextCallback) : FunctionReturn {
		let result: FunctionReturn;
		
		// do something before calling the next interceptor or the final implementation
		
		result = getNext()(input, getNext);
		
		// do something after calling the next interceptor or the final implementation
		
		return result;
	}
}
```

Interceptor registration based in decorators
```javascript
@Intercept({
	behavior: MyInterceptionBehavior,
	order: 1
})
class MyService {
	@Intercept({
		behavior: MyInterceptionBehavior,
		order: 2
	})
	doSomething(): void {
	}
}
```

And also manual interceptor registration for types which are out of my control, e.g. Router or Http
```javascript
bind(MyService).toFactory(() => ....).interceptWith(MyInterceptor)
```

The interception API from Unity has one drawback. It does not support interface interception without a target implementation. But this would also be nice to have. Imagine I have several service interface where I can provide a generic implementation without to be forced implementing every single interface in the same way. An example usage could be a service proxy. May be this could look like following:

```javascript
bind(MyService).toInterceptor(MyInterfaceInterceptor)
```

**Comment by robwormald** [comment]
Tracking this idea over in #2684

**Comment by jeffbcross** [comment]
@lehmamic this sounds more like the provider API in Angular 1 than the $http interceptors API, wouldn't you say?

**Comment by lehmamic** [comment]
@jeffbcross I would rather compare it to the connect module of Node JS than with the provider of $http interceptors API in Angular 1.

The interceptor API i'm proposing would create a proxy around the service. The proxy has a pipeline of interceptors (similar to the middleware pipeline in Connect) which gets executed during every method and property call of the service. By calling `result = getNext()(input, getNext);` the interceptor will invoke the next interceptor in the pipeline or, in case it is the last interceptor, the original service function.
![Interceptor Pipeline](https://img5.picload.org/image/iprwiir/interceptionpipeline.gif)

An interceptor would not get fix defined parameter like req, resp or $http, but the metadata of the called service and method. This would include information about the parameter types, values return types, values etc. I called this FunctionInvocation in my sample above.

This would allow us, to attach behaviors (which have to be written only once) to any service (not only $http), which is very powerful. I consider this as very valuable for large enterprise applications.

**Comment by mhevery** [comment]
This is obsolete, and no longer relevant / actionable. To keep our issues clean, we are aggressively closing them.

**Comment by squadwuschel** [comment]
When this is obsolete, what is the solution for this kind of interception problem, it would be cool when you would explain why such a great problem is obsolete.

**Comment by Delagen** [comment]
@squadwuschel You can write a wrapper on your own Service. With RxJS it's make easy to do this pipelines.

**Comment by squadwuschel** [comment]
I've written my own http provider by extending the http but the problem is that you can only register one http provider in your ngmodule but I have around 6 different implementations and I can't implement them all in one place and I was not able to inject ui-router into the extended http. So you see problems over Problems and I can't find a good solution for this.

**Comment by mhevery** [comment]
RxJS has a lot of facilities for such interception.

**Comment by nolazybits** [comment]
I realise it is bring it back fromt he dead but @mhevery any example then would be great please.
Thanks

**Comment by mhevery** [comment]
@alxhub is working on a new proposal which will be available after he is done.

**Comment by hikumealan** [comment]
@mhevery where can we track the progress @alxhub is making on this request?
Interception is a great feature which would be great to see apart of standard Angular 2.

**Comment by DzmitryShylovich** [comment]
https://github.com/angular/angular/pull/14334
